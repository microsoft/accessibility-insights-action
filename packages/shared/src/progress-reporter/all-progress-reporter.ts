// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';
import { ProgressReporter } from './progress-reporter';
import { CombinedReportParameters } from 'accessibility-insights-report';
import { iocTypes } from '../ioc/ioc-types';
import { BaselineEvaluation } from 'accessibility-insights-scan';
import { throwOnAnyErrors } from '../utils/aggregate-error';

@injectable()
export class AllProgressReporter extends ProgressReporter {
    constructor(@inject(iocTypes.ProgressReporters) protected readonly progressReporters: ProgressReporter[]) {
        super();
    }

    public async start(): Promise<void> {
        await this.execute((r) => r.start());
    }

    public async completeRun(combinedReportResult: CombinedReportParameters, baselineEvaluation?: BaselineEvaluation): Promise<void> {
        if (baselineEvaluation) {
            await this.execute((r) => r.completeRun(combinedReportResult, baselineEvaluation));
        } else {
            await this.execute((r) => r.completeRun(combinedReportResult));
        }
    }

    public async failRun(): Promise<void> {
        await this.execute((r) => r.failRun());
    }

    public async didScanSucceed(): Promise<boolean> {
        let allReportersSucceeded = true;

        await this.execute(async (r) => {
            if (!(await r.didScanSucceed())) {
                allReportersSucceeded = false;
            }
        });

        return Promise.resolve(allReportersSucceeded);
    }

    private async execute(callback: (reporter: ProgressReporter) => Promise<void>): Promise<void> {
        const errors = [];
        const length = this.progressReporters.length;
        for (let pos = 0; pos < length; pos += 1) {
            try {
                await callback(this.progressReporters[pos]);
            } catch (e) {
                if (e instanceof Error) {
                    errors.push(e);
                } else {
                    throw e;
                }
            }
        }

        throwOnAnyErrors(errors, 'Multiple progress reporters encountered Errors');
    }
}
