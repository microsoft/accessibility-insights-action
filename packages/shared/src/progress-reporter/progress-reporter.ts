// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable } from 'inversify';
import { CombinedReportParameters } from 'accessibility-insights-report';
import { Renderer, marked } from 'marked';
import { BaselineEvaluation } from 'accessibility-insights-scan';

@injectable()
export abstract class ProgressReporter {
    constructor() {
        if (process.env.ACT === 'true') {
            marked.setOptions({
                renderer: new Renderer(),
            });
        }
    }

    public abstract start(): Promise<void>;
    public abstract completeRun(combinedReportResult: CombinedReportParameters, baselineEvaluation?: BaselineEvaluation): Promise<void>;
    public abstract failRun(): Promise<void>;
    public didScanSucceed(): Promise<boolean> {
        return Promise.resolve(true); // Will be overridden in classes that use this
    }

    protected async invoke<T>(fn: () => Promise<T>): Promise<T> {
        return process.env.ACT !== 'true' ? fn() : Promise.resolve(undefined as T);
    }

    protected traceMarkdown(markdown: string): void {
        if (process.env.ACT === 'true') {
            console.log('[ProgressReporter] ===');
            console.log(marked(markdown));
        }
    }
}
