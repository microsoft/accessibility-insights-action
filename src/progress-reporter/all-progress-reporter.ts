// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';
import { CheckRunCreator } from './check-run/check-run-creator';
import { ProgressReporter } from './progress-reporter';
import { PullRequestCommentCreator } from './pull-request/pull-request-comment-creator';
import { CombinedReportParameters } from 'accessibility-insights-report';

@injectable()
export class AllProgressReporter extends ProgressReporter {
    private readonly progressReporters: ProgressReporter[];

    constructor(
        @inject(PullRequestCommentCreator) pullRequestCommentCreator: PullRequestCommentCreator,
        @inject(CheckRunCreator) checkRunCreator: CheckRunCreator,
    ) {
        super();
        this.progressReporters = [checkRunCreator, pullRequestCommentCreator];
    }

    public async start(): Promise<void> {
        await this.execute((r) => r.start());
    }

    public async completeRun(combinedReportResult: CombinedReportParameters): Promise<void> {
        await this.execute((r) => r.completeRun(combinedReportResult));
    }

    public async failRun(message: string): Promise<void> {
        await this.execute((r) => r.failRun(message));
    }

    private async execute(callback: (reporter: ProgressReporter) => Promise<void>): Promise<void> {
        const length = this.progressReporters.length;
        for (let pos = 0; pos < length; pos += 1) {
            await callback(this.progressReporters[pos]);
        }
    }
}
