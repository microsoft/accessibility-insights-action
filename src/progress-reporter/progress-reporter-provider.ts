// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as github from '@actions/github';
import { Octokit } from '@octokit/rest';
import { inject, injectable } from 'inversify';

import { iocTypes } from '../ioc/ioc-types';
import { Logger } from '../logger/logger';
import { CheckRunCreator } from './check-run/check-run-creator';
import { ProgressReporter } from './progress-reporter';
import { PullRequestCommentCreator } from './pull-request/pull-request-comment-creator';

@injectable()
export class ProgressReporterProvider {
    constructor(
        @inject(PullRequestCommentCreator) private readonly pullRequestCommentCreator: PullRequestCommentCreator,
        @inject(CheckRunCreator) private readonly checkRunCreator: CheckRunCreator,
        @inject(iocTypes.Github) private readonly githubObj: typeof github,
        @inject(Logger) private readonly logger: Logger,
    ) {}

    public getInstance(): ProgressReporter {
        if (this.isPullRequest()) {
            return this.pullRequestCommentCreator;
        }

        return this.checkRunCreator;
    }

    private isPullRequest(): boolean {
        this.logMessage(`event name - ${this.githubObj.context.eventName}`);

        return this.githubObj.context.eventName === 'pull_request';
    }

    private logMessage(message: string): void {
        this.logger.logInfo(`[ProgressReporterProvider] ${message}`);
    }
}
