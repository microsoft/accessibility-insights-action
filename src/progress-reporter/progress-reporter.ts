// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable } from 'inversify';
import { CombinedReportParameters } from 'accessibility-insights-report';
import * as marked from 'marked';
import * as TerminalRenderer from 'marked-terminal';

@injectable()
export abstract class ProgressReporter {
    constructor() {
        if (process.env.ACT === 'true') {
            marked.setOptions({
                renderer: new TerminalRenderer(),
            });
        }
    }

    public abstract start(): Promise<void>;
    public abstract completeRun(combinedReportResult: CombinedReportParameters): Promise<void>;
    public abstract failRun(message: string): Promise<void>;

    protected async invoke<T>(fn: () => Promise<T>): Promise<T> {
        return process.env.ACT !== 'true' ? fn() : Promise.resolve(undefined as T);
    }

    protected traceMarkdown(markdown: string): void {
        if (process.env.ACT === 'true') {
            console.log('[ProgressReporter] ===');
            console.log(marked(markdown));
        }
    }
}
