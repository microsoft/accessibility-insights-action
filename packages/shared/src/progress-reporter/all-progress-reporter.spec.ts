// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { CombinedReportParameters } from 'accessibility-insights-report';
import 'reflect-metadata';

import { IMock, Mock, Times } from 'typemoq';
import { BaselineEvaluation } from 'accessibility-insights-scan';
import { AllProgressReporter } from './all-progress-reporter';
import { ProgressReporter } from './progress-reporter';

describe(AllProgressReporter, () => {
    let testSubject: AllProgressReporter;
    let progressReporterMock: IMock<ProgressReporter>;

    beforeEach(() => {
        progressReporterMock = Mock.ofType<ProgressReporter>();
        testSubject = new AllProgressReporter([progressReporterMock.object]);
    });

    it('start should invoke all reporters', async () => {
        executeOnReporter((reporter) => {
            reporter
                .setup((p) => p.start())
                .returns(() => Promise.resolve())
                .verifiable(Times.once());
        });

        await testSubject.start();
    });

    describe('complete', () => {
        it('should invoke all reporters', async () => {
            const axeResults = 'axe results' as unknown as CombinedReportParameters;
            executeOnReporter((reporter) => {
                reporter
                    .setup((p) => p.completeRun(axeResults))
                    .returns(() => Promise.resolve())
                    .verifiable(Times.once());
            });

            await testSubject.completeRun(axeResults);
        });

        it('should invoke all reporters when baselineEvaluation is available', async () => {
            const axeResults = 'axe results' as unknown as CombinedReportParameters;
            const baselineEvalStub = 'baseline evaluation' as unknown as BaselineEvaluation;
            executeOnReporter((reporter) => {
                reporter
                    .setup((p) => p.completeRun(axeResults, baselineEvalStub))
                    .returns(() => Promise.resolve())
                    .verifiable(Times.once());
            });

            await testSubject.completeRun(axeResults, baselineEvalStub);
        });
    });

    it('failRun should invoke all reporters', async () => {
        const error = 'scan error';
        executeOnReporter((reporter) => {
            reporter
                .setup((p) => p.failRun(error))
                .returns(() => Promise.resolve())
                .verifiable(Times.once());
        });

        await testSubject.failRun(error);
    });

    afterEach(() => {
        progressReporterMock.verifyAll();
    });

    function executeOnReporter(callback: (reporter: IMock<ProgressReporter>) => void): void {
        callback(progressReporterMock);
    }
});
