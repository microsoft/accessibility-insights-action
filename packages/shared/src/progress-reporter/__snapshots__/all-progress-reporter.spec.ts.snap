// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllProgressReporter complete should rethrow an AggregateError if multiple reporters throw 1`] = `
"Multiple progress reporters encountered Errors
    error from failingReporter
    error from failingReporter"
`;

exports[`AllProgressReporter didScanSucceed should rethrow an AggregateError if multiple reporters throw 1`] = `
"Multiple progress reporters encountered Errors
    error from failingReporter
    error from failingReporter"
`;

exports[`AllProgressReporter failRun should rethrow an AggregateError if multiple reporters throw 1`] = `
"Multiple progress reporters encountered Errors
    error from failingReporter
    error from failingReporter"
`;

exports[`AllProgressReporter start should rethrow an AggregateError if multiple reporters throw 1`] = `
"Multiple progress reporters encountered Errors
    error from failingReporter
    error from failingReporter"
`;
